src/core/inductive.ml

```diff
@@ -0,0 +1,280 @@
+(** Generating of induction principles. *)
+
+open Timed
+open Pos
+open Console
+open Terms
+open Print
+open Syntax
+
+(** Logging function for generating of inductive principle. *)
+let log_ind = new_logger 'g' "indu" "generating induction principle"
+let log_ind = log_ind.logger
+
+(** Builtin configuration for induction. *)
+type config =
+  { symb_Prop : sym (** : TYPE. Type of propositions. *)
+  ; symb_prf  : sym (** : Prop → TYPE.
+                        Interpretation of propositions as types. *) }
+
+(** [get_config ss pos] build the configuration using [ss]. *)
+let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
+  let builtin = Builtin.get ss pos in
+  { symb_Prop = builtin "Prop"
+  ; symb_prf  = builtin "P" }
+
+(** [unbind_name b] unbinds the binder [b] and substitutes the variable in the
+    case of undependent product. It returns the variable and the term which
+    was the body of the binder. *)
+let unbind_name : (term, term) Bindlib.binder -> tvar * term = fun b ->
+  if Bindlib.binder_occur b then
+    Bindlib.unbind b
+  else
+    let x = Bindlib.new_var mkfree "x" in
+    (x, Bindlib.subst b (Vari x))
+
+(** [create_instance_of pos sind f] creates an instance of the symbol [sind],
+    which defined on the position [pos], and applies [f] to this instance.
+    It's useful to build the type of the property p or build the conclusion
+    of the inductive principle. *)
+let create_instance_of : popt -> sym -> (tbox list -> tbox) -> tbox =
```

Comment:
create_instance_of is a name that is too generic. I propose to use a more specific name, say change_codom_ind_typ.
```
(** [change_codom_ind_typ pos ind_sym codom] assumes that the type of [ind_sym] is of the form [\Pi x1:a1, .., \Pi xn:an, TYPE]. It then generates a [tbox] similar to this type except that [TYPE] is replaced by [codom [x1;..;xn]]. *)
let change_codom_ind_typ : popt -> sym -> (tbox list -> tbox) -> tbox =
fun pos ind_sym f ->
  let rec aux : tvar list -> term -> tbox = fun xs a ->
    match Basics.get_args a with
    | (Type, _) -> codom (List.rev_map _Vari xs)
    | (Prod(a,b), _) ->
        let (x,b) = unbind_name b in
        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
    | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
  in aux [] !(ind_sym.sym_type)
```